triedb/pathdb/nodebuffer.go

```diff
@@ -212,35 +214,45 @@ func (b *nodebuffer) allocBatch(db ethdb.KeyValueStore) ethdb.Batch {
 	return db.NewBatchWithSize((metasize + int(b.size)) * 11 / 10) // extra 10% for potential pebble internal stuff
 }
 
-// flush persists the in-memory dirty trie node into the disk if the configured
-// memory threshold is reached. Note, all data must be written atomically.
-func (b *nodebuffer) flush(db ethdb.KeyValueStore, clean *fastcache.Cache, id uint64, force bool) error {
-	if b.size <= b.limit && !force {
-		return nil
-	}
-	// Ensure the target state id is aligned with the internal counter.
-	head := rawdb.ReadPersistentStateID(db)
-	if head+b.layers != id {
-		return fmt.Errorf("buffer layers (%d) cannot be applied on top of persisted state id (%d) to reach requested state id (%d)", b.layers, head, id)
-	}
-	var (
-		start = time.Now()
-		batch = b.allocBatch(db)
-	)
-	nodes := writeNodes(batch, b.nodes, clean)
-	rawdb.WritePersistentStateID(batch, id)
+// flush persists the in-memory dirty trie node into the disk. Note, all data must be written atomically.
+func (b *nodebuffer) flush(db ethdb.KeyValueStore, clean *fastcache.Cache, id uint64) {
+	b.done = make(chan struct{})
 
-	// Flush all mutations in a single batch
-	size := batch.ValueSize()
-	if err := batch.Write(); err != nil {
-		return err
-	}
-	commitBytesMeter.Mark(int64(size))
-	commitNodesMeter.Mark(int64(nodes))
-	commitTimeTimer.UpdateSince(start)
-	log.Debug("Persisted pathdb nodes", "nodes", len(b.nodes), "bytes", common.StorageSize(size), "elapsed", common.PrettyDuration(time.Since(start)))
-	b.reset()
-	return nil
+	go func() {
+		defer func() {
+			close(b.done)
+		}()
+		// Error out if the state id is aligned with the internal counter
+		head := rawdb.ReadPersistentStateID(db)
+		if head+b.layers != id {
+			b.flushErr = fmt.Errorf("buffer layers (%d) cannot be applied on top of persisted state id (%d) to reach requested state id (%d)", b.layers, head, id)
+			return
+		}
+		var (
+			start = time.Now()
+			batch = b.allocBatch(db)
+		)
+		nodes := writeNodes(batch, b.nodes, clean)
+		rawdb.WritePersistentStateID(batch, id)
+
+		// Flush all mutations in a single batch
```

Comment:
How about this
in `disklayer.go`, function `node()`, we lookup a node. Order: 

1. buffer 
2. frozen
3. cleans
4. database
   - And if found, write to cleans

```golang
	if dl.cleans != nil && len(blob) > 0 {
		dl.cleans.Set(key, blob)
		cleanWriteMeter.Mark(int64(len(blob)))
	}
```
I'm trying to think of a case where this write-to-cleans conflicts with the write-to-cleans in the background committer `writeNodes` method.

Reply:
- if it's found in buffer/frozen => return and no interaction with cache
- if it's found in cache => return
- if it's found in disk (it implicitly means the item is not in these places above, even the item is marked as deleted, it will be caught in either buffer/frozen/cache), load it from db and add it into the cache

so, no conflict should happen

But i have to say it's a really good point, i haven't thought about it

Reply:
> Regarding "flush all mutations in a single batch" -- is that important only because of crash-safety, or some other more subtle reason?

Only because of crash-safety